### PR TITLE
Hibernate MUC processes

### PR DESCRIPTION
**Pull request by michalwski**
This PR hibernates the MUC process after configurable period of inactivity. This is very similar to what we already have for c2s process.

The timeout is configurable and default value is 90s.

### CODE REVIEW

File: apps/ejabberd/src/mod_muc_room.erl [hunk 1]
```diff
@@ -187,7 +187,6 @@ start_link(Host, ServerHost, Access, Room, HistorySize, RoomShaper, HttpAuthPool
                         RoomShaper, HttpAuthPool, Opts],
                        ?FSMOPTS).
 
-
 -spec get_room_users(RoomJID :: ejabberd:jid()) -> {ok, [#user{}]}
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> The spec in line 190 uses a record, please define a type for the record and use that instead.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 2]
```diff
@@ -3781,6 +3783,40 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
     end).
 
+room_is_hibernated(Config) ->
+    RoomName = fresh_room_name(),
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        Username = escalus_client:username(Alice),
+        Server = escalus_client:server(Alice),
+        Resource = escalus_client:resource(Alice),
+        JID = {jid, Username, Server, Resource, escalus_utils:jid_to_lower(Username), Server, Resource},
+        RoomJID = {jid, RoomName, muc_host(), <<>>, escalus_utils:jid_to_lower(RoomName), muc_host(), <<>>},
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3793 is too long:         RoomJID = {jid, RoomName, muc_host(), <<>>, escalus_utils:jid_to_lower(RoomName), muc_host(), <<>>},.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 3]
```diff
@@ -628,9 +632,9 @@ handle_info({process_user_presence, From}, normal_state = _StateName, StateData)
     if
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Replace the 'if' expression on line 632 with a 'case' expression or function clauses.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 4]
```diff
@@ -639,9 +643,9 @@ handle_info({process_user_message, From}, normal_state = _StateName, StateData)
...
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Replace the 'if' expression on line 643 with a 'case' expression or function clauses.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 5]
```diff
@@ -398,7 +400,7 @@ locked_state({route, From, _ToNick,
         locked_state ->
             {next_state, NextState2, StateData3};
         normal_state ->
-            {next_state, NextState2, StateData3#state{just_created = false}}
+            {next_state, NextState2, StateData3#state{just_created = false}, StateData3#state.hibernate_timeout}
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Line 403 is too long:             {next_state, NextState2, StateData3#state{just_created = false}, StateData3#state.hibernate_timeout}.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 6]
```diff
@@ -3781,6 +3783,40 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
         escalus_assert:has_no_stanzas(Alice)
     end).
 
+room_is_hibernated(Config) ->
+    RoomName = fresh_room_name(),
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        Username = escalus_client:username(Alice),
+        Server = escalus_client:server(Alice),
+        Resource = escalus_client:resource(Alice),
+        JID = {jid, Username, Server, Resource, escalus_utils:jid_to_lower(Username), Server, Resource},
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3792 is too long:         JID = {jid, Username, Server, Resource, escalus_utils:jid_to_lower(Username), Server, Resource},.

File: apps/ejabberd/src/mod_muc.erl [hunk 7]
```diff
@@ -140,6 +144,7 @@ start_link(Host, Opts) ->
 -spec start(ejabberd:server(),_) -> {'error',_}
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Missing space right "," on line 144

File: apps/ejabberd/src/mod_muc.erl [hunk 8]
```diff
@@ -140,6 +144,7 @@ start_link(Host, Opts) ->
...
             | {'ok','undefined' | pid()} | {'ok','undefined' | pid(),_}.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Missing space right "," on line 145

File: apps/ejabberd/src/mod_muc.erl [hunk 9]
```diff
@@ -1127,3 +1132,52 @@ can_access_room(_, From, To) ->
...
+
+count_hibernated_rooms(Pid, Count) ->
+    case erlang:process_info(Pid, current_function) of
+        {current_function, {erlang, hibernate, _}} ->
+            Count + 1;
+        _ ->
+            Count
+    end.
+
+-define(EX_EVAL_SINGLE_VALUE, {[{l, [{t, [value, {v, 'Value'}]}]}],[value]}).
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Missing space right "," on line 1174

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 10]
```diff
@@ -3781,6 +3797,133 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+        given_fresh_room_with_messages_is_hibernated(Alice, RoomName, [], Bob)
+
+    end),
+
+    destroy_room(muc_host(), RoomName).
+
+hibernated_room_can_be_queried_for_archive(Config) ->
+    RoomName = fresh_room_name(),
+    escalus:fresh_story(Config, [{alice, 1}, {bob, 1}], fun(Alice, Bob) ->
+        {Msg, {ok, _, Pid}} = given_fresh_room_with_messages_is_hibernated(Alice, RoomName, [], Bob),
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3844 is too long:         {Msg, {ok, _, Pid}} = given_fresh_room_with_messages_is_hibernated(Alice, RoomName, [], Bob),.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 11]
```diff
@@ -3781,6 +3797,133 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+    {ok, _, RoomPid} = Result = given_fresh_room_for_user(Owner, RoomName, Opts),
+    true = wait_for_hibernation(RoomPid, 10),
+    Result.
+
+given_fresh_room_for_user(Owner, RoomName, Opts) ->
+    Username = escalus_client:username(Owner),
+    Server = escalus_client:server(Owner),
+    Resource = escalus_client:resource(Owner),
+    JID = {jid, Username, Server, Resource, escalus_utils:jid_to_lower(Username), Server, Resource},
+    RoomJID = {jid, RoomName, muc_host(), <<>>, escalus_utils:jid_to_lower(RoomName), muc_host(), <<>>},
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3872 is too long:     RoomJID = {jid, RoomName, muc_host(), <<>>, escalus_utils:jid_to_lower(RoomName), muc_host(), <<>>},.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 12]
```diff
@@ -628,9 +636,9 @@ handle_info({process_user_presence, From}, normal_state = _StateName, StateData)
     if
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Replace the 'if' expression on line 636 with a 'case' expression or function clauses.

File: apps/ejabberd/src/mod_muc.erl (hunk 8 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Missing space right "," on line 145

File: apps/ejabberd/src/mod_muc_room.erl [hunk 13]
```diff
@@ -639,9 +647,9 @@ handle_info({process_user_message, From}, normal_state = _StateName, StateData)
...
     RoomQueueEmpty ->
         StateData2 = prepare_room_queue(StateData1),
-        {next_state, normal_state, StateData2};
+        next_normal_state(StateData2);
     true ->
-        {next_state, normal_state, StateData1}
+        next_normal_state(StateData1)
     end;
 handle_info(process_room_queue, normal_state = StateName, StateData) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Compiler**:
> Warning: variable 'StateName' is unused

File: apps/ejabberd/src/mod_muc_room.erl [hunk 14]
```diff
@@ -516,14 +518,17 @@ normal_state({route, From, ToNick,
...
         JIDs -> lists:foreach(FunRouteNickIq, JIDs)
     end,
-    {next_state, normal_state, StateData};
+    next_normal_state(StateData);
 normal_state({http_auth, AuthPid, Result, From, Nick, Packet, Role}, StateData) ->
     AuthPids = StateData#state.http_auth_pids,
     StateDataWithoutPid = StateData#state{http_auth_pids = lists:delete(AuthPid, AuthPids)},
     NewStateData = handle_http_auth_result(Result, From, Nick, Packet, Role, StateDataWithoutPid),
     destroy_temporary_room_if_empty(NewStateData, normal_state);
+normal_state(timeout, #state{server_host = Host} = StateData) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Compiler**:
> Warning: variable 'Host' is unused

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 15]
```diff
@@ -3781,6 +3813,133 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3888 is too long:     RoomJID = {jid, RoomName, muc_host(), <<>>, escalus_utils:jid_to_lower(RoomName), muc_host(), <<>>},.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 16]
```diff
@@ -3781,6 +3813,133 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+    destroy_room(muc_host(), RoomName).
+
+hibernation_metrics_are_updated(Config) ->
+    RoomName = fresh_room_name(),
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        given_fresh_room_is_hibernated(Alice, RoomName, []),
+
+        OnlineRooms = escalus_ejabberd:rpc(mod_muc, online_rooms_number, []),
+        true = OnlineRooms > 0,
+        Hibernations = escalus_ejabberd:rpc(mongoose_metrics, get_metric_value, [global, [mod_muc, hibernations]]),
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3839 is too long:         Hibernations = escalus_ejabberd:rpc(mongoose_metrics, get_metric_value, [global, [mod_muc, hibernations]]),.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 17]
```diff
@@ -639,9 +647,9 @@ handle_info({process_user_message, From}, normal_state = _StateName, StateData)
     if
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Replace the 'if' expression on line 647 with a 'case' expression or function clauses.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 18]
```diff
@@ -3781,6 +3813,133 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+        given_fresh_room_with_messages_is_hibernated(Alice, RoomName, [], Bob)
+
+    end),
+
+    destroy_room(muc_host(), RoomName).
+
+hibernated_room_can_be_queried_for_archive(Config) ->
+    RoomName = fresh_room_name(),
+    escalus:fresh_story(Config, [{alice, 1}, {bob, 1}], fun(Alice, Bob) ->
+        {Msg, {ok, _, Pid}} = given_fresh_room_with_messages_is_hibernated(Alice, RoomName, [], Bob),
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3860 is too long:         {Msg, {ok, _, Pid}} = given_fresh_room_with_messages_is_hibernated(Alice, RoomName, [], Bob),.

File: apps/ejabberd/src/mod_muc.erl [hunk 19]
```diff
@@ -1127,3 +1132,52 @@ can_access_room(_, From, To) ->
...
+    end.
+
+hibernated_rooms_number() ->
+    lists:sum([hibernated_rooms_number(Host) || Host <- ?MYHOSTS]).
+
+hibernated_rooms_number(Host) ->
+    try
+        count_hibernated_rooms(Host)
+    catch _:_ ->
+              0
```
Comment (RoXeon) [inline apps/ejabberd/src/mod_muc.erl]: This indentation seems wired. Is it valid in terms of code-style of this project? Sorry about maybe redundant question, I'm trying to figure out exactly how the code-style looks in this project to improve my future code :)

Comment (michalwski) [inline apps/ejabberd/src/mod_muc.erl]: In this particular case I just followed vim's indentation which is in most cases what we want.

**Pull request merged**